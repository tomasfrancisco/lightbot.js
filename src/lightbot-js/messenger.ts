import { v4 as uuid } from "uuid";

import { LightbotAPI } from "./api";
import { APIMessage } from "./api.types";
import { StateManager } from "./state-manager";

export type LightbotMessage = APIMessage & {
  sender: "bot" | "human";
};

export type Message = APIMessage;

export type OnChangeHandler = () => void;

export interface LightbotMessengerProps {
  hostURL: string;
  agentId: string;
  onChange?: OnChangeHandler;
}

export class LightbotMessenger {
  private stateManager: StateManager;
  private apiClient: LightbotAPI;
  private onChange?: OnChangeHandler;

  constructor({ hostURL, agentId, onChange }: LightbotMessengerProps) {
    this.stateManager = new StateManager();
    this.apiClient = new LightbotAPI(
      hostURL,
      agentId,
      // TODO: Remove when backend gets rid of session and user ids
      uuid(),
      // TODO: Remove when backend gets rid of session and user ids
      uuid(),
    );
    if (onChange) {
      this.onChange = onChange;
    }

    this.initAgent();
  }

  /**
   * It will set the UI messenger state (Open or Close)
   */
  public toggleMessenger = async (): Promise<void> => {
    this.stateManager.updateLayout({
      isMessengerOpen: !this.stateManager.layout.isMessengerOpen,
    });

    this.pushUpdate();
  };

  public get isMessengerOpen(): boolean {
    return !!this.stateManager.layout.isMessengerOpen;
  }

  public get messages(): LightbotMessage[] {
    return this.stateManager.messages;
  }

  public sendMessage = async (message: Message): Promise<void> => {
    if (!this.stateManager.agent.isInitialized) {
      throw new Error(
        `
          Lightbot messenger was not initialized.
          Please refresh the page or make sure LightbotMessenger instance is correctly initialized.
        `,
      );
    }

    this.stateManager.saveMessages([{ ...message, sender: "human" }], this.pushUpdate);

    try {
      let messagesResponse: APIMessage[] | undefined;
      if (message.type === "jump") {
        messagesResponse = await this.apiClient.postJump(message.label);
      } else {
        messagesResponse = await this.apiClient.postMessage(message.label);
      }

      if (messagesResponse) {
        const messages: LightbotMessage[] = messagesResponse.map<LightbotMessage>(
          messageResponse => ({
            ...messageResponse,
            sender: "bot",
          }),
        );

        this.stateManager.saveMessages(messages, this.pushUpdate);
      }
    } catch (err) {
      this.stateManager.popMessage(this.pushUpdate);
      throw new Error("An error occurred sending a message.");
    }
  };

  public resetAgent = () => {
    this.stateManager.resetState();
    this.pushUpdate();
  };

  /**
   * Triggers an opening message from the bot.
   */
  private initAgent = async () => {
    if (this.stateManager.agent.isInitialized) {
      console.warn("Lightbot messenger was already initialized.");
      return;
    }

    try {
      const messagesResponse = await this.apiClient.postStartMessenger();
      if (messagesResponse) {
        const messages: LightbotMessage[] = messagesResponse.map<LightbotMessage>(message => ({
          ...message,
          sender: "bot",
        }));

        this.stateManager.saveMessages(messages, this.pushUpdate);
      }

      const agentData = await this.apiClient.getAgentData();
      if (agentData && typeof agentData === "object") {
        this.stateManager.updateAgent(agentData);
      }

      await this.stateManager.updateAgent({ isInitialized: true });
    } catch (err) {
      console.warn("An error occurred initializing messenger.");
    }

    await this.stateManager.updateAgent({ isInitialized: true });

    this.pushUpdate();
  };

  /**
   * Notifies the subscriber with the updated messages
   */
  private pushUpdate = () => {
    if (this.onChange) {
      setTimeout(this.onChange, 0);
    }
  };
}
