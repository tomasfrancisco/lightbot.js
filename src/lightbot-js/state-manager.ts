import { APIAgentData } from "lightbot-js/api.types";

import { LightbotMessage } from "./messenger";

export interface LayoutState {
  isMessengerOpen?: boolean;
  [propName: string]: any;
}

export type AgentState = APIAgentData & {
  isInitialized?: boolean;
  [propName: string]: any;
};

/**
 * Internal State for State Manager
 * Some properties are not exposed, so it means the type should not be exposed either
 */
interface StoreState {
  isLocalStorageAvailable?: boolean;
  messages: LightbotMessage[];
  agent: AgentState;
  layout: LayoutState;
}

type StoreKeys = { [K in keyof StoreState]: string };

/**
 * State Manager
 * is responsible for keeping messages and bot agent relative data.
 * Non critical information like messages and agent data is stored in browser storage.
 *
 * Session and User data should be kept under the control of the API - e.g. cookies.
 * WARNING: NEVER USE STATE MANAGER TO DEAL WITH CRITICAL DATA
 */
export class StateManager {
  public static keys: StoreKeys = {
    agent: StateManager.getKey("agent"),
    layout: StateManager.getKey("layout"),
    messages: StateManager.getKey("messages"),
  };
  private static getKey(value: keyof StoreState) {
    return `"19Hgw012xn!@"${value}`;
  }

  private state: StoreState = {
    agent: {
      isInitialized: false,
    },
    layout: {
      isMessengerOpen: false,
    },
    messages: [],
  };

  constructor() {
    this.initMessagesState();
    this.initLayoutState();
    this.initAgentState();
  }

  public saveMessages(messages: LightbotMessage[], callback?: () => void) {
    this.state.messages = this.state.messages.concat(messages);
    this.updateMessagesStorage();
    if (callback) {
      callback();
    }
  }

  public popMessage(callback?: () => void) {
    const message = this.state.messages.pop();
    this.updateMessagesStorage();
    if (callback) {
      callback();
    }
    return message;
  }

  public get messages() {
    return this.state.messages;
  }

  /**
   * Interface to override existing or create new layout property values
   * @param layout Overrides to layout store
   */
  public updateLayout(layout: LayoutState) {
    this.state.layout = Object.assign({}, this.state.layout, layout);
    this.updateLayoutStorage();
  }

  public get layout() {
    return this.state.layout;
  }

  public updateAgent(agent: AgentState) {
    this.state.agent = Object.assign({}, this.state.agent, agent);
    this.updateAgentStorage();
  }

  public get agent() {
    return this.state.agent;
  }

  private initMessagesState() {
    if (!this.testLocalStorage()) {
      console.warn(
        "An error occurred while initalizing message storage. LocalStorage is not available",
      );
      return;
    }

    const messages = localStorage.getItem(StateManager.keys.messages);

    if (!messages) {
      localStorage.setItem(StateManager.keys.messages, this.serialize(this.state.messages));
    } else {
      try {
        this.state.messages = this.deserialize(messages);
      } catch (err) {
        localStorage.setItem(StateManager.keys.messages, this.serialize(this.state.messages));
        console.warn(
          "An error occurred while initalizing message storage. The message history was reset",
        );
      }
    }
  }

  private updateMessagesStorage() {
    if (!this.testLocalStorage()) {
      return;
    }

    const messages = localStorage.getItem(StateManager.keys.messages);

    if (!messages) {
      localStorage.setItem(StateManager.keys.messages, this.serialize(this.state.messages));
      console.warn(
        "An error occurred while removing last message. The messages history was reset.",
      );
      return;
    }

    localStorage.setItem(StateManager.keys.messages, this.serialize(this.state.messages));
  }

  private initLayoutState() {
    if (!this.testLocalStorage()) {
      console.warn(
        "An error occurred while initalizing layout storage. LocalStorage is not available.",
      );
      return;
    }

    const layout = localStorage.getItem(StateManager.keys.layout);

    if (!layout) {
      localStorage.setItem(StateManager.keys.layout, this.serialize(this.state.layout));
    } else {
      try {
        this.state.layout = this.deserialize(layout);
      } catch (err) {
        localStorage.setItem(StateManager.keys.layout, this.serialize(this.state.layout));
        console.warn("An error occurred while initalizing layout storage. The layout was reset.");
      }
    }
  }

  private updateLayoutStorage() {
    if (!this.testLocalStorage()) {
      return;
    }

    const layout = localStorage.getItem(StateManager.keys.layout);

    if (!layout) {
      localStorage.setItem(StateManager.keys.layout, this.serialize(this.state.layout));
      console.warn("An error occurred while removing last message. The layout was reset.");
    }

    localStorage.setItem(StateManager.keys.layout, this.serialize(this.state.layout));
  }

  private initAgentState() {
    if (!this.testLocalStorage()) {
      console.warn(
        "An error occurred while initalizing agent storage. LocalStorage is not available.",
      );
      return;
    }

    const agent = localStorage.getItem(StateManager.keys.agent);

    if (!agent) {
      localStorage.setItem(StateManager.keys.agent, this.serialize(this.state.agent));
    } else {
      try {
        this.state.agent = this.deserialize(agent);
      } catch (err) {
        localStorage.setItem(StateManager.keys.agent, this.serialize(this.state.agent));
        console.warn("An error occurred while initalizing agent storage. The agent was reset");
      }
    }
  }

  private updateAgentStorage() {
    if (!this.testLocalStorage()) {
      return;
    }

    const agent = localStorage.getItem(StateManager.keys.agent);

    if (!agent) {
      localStorage.setItem(StateManager.keys.agent, this.serialize([]));
      throw new Error("An error occurred while updating agent. The agent was reset.");
    }

    localStorage.setItem(StateManager.keys.agent, this.serialize(this.state.agent));
  }

  /**
   * Returns true if LocalStorage is available
   */
  private testLocalStorage() {
    if (this.state.isLocalStorageAvailable !== undefined) {
      return this.state.isLocalStorageAvailable;
    }

    const testKey = "localstorage_test_key";
    const testValue = "localstorage_test_value";
    try {
      localStorage.setItem(testKey, testValue);
      if (localStorage.getItem(testKey) === testValue) {
        localStorage.removeItem(testKey);
        this.state.isLocalStorageAvailable = true;
      }
    } catch (err) {
      this.state.isLocalStorageAvailable = false;
    }
    return this.state.isLocalStorageAvailable;
  }

  private serialize(value: any) {
    return JSON.stringify(value);
  }

  private deserialize(value: any) {
    return JSON.parse(value);
  }
}
