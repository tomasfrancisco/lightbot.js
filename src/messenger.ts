import { LightbotAPI } from "./api";
import { APIMessage } from "./api.types";
import { StateManager } from "./state-manager";

export type Message = APIMessage & {
  sender: "human" | "bot" | "supporter";
};

export type UpdateListenerHandler = () => void;

export interface LightbotMessengerProps {
  hostURL: string;
  agentId: string;
  updateListener?: UpdateListenerHandler;
}

export class LightbotMessenger {
  private stateManager: StateManager;
  private apiClient: LightbotAPI;
  private updateListener?: UpdateListenerHandler;

  constructor({ hostURL, agentId, updateListener: messageListener }: LightbotMessengerProps) {
    this.stateManager = new StateManager();
    this.apiClient = new LightbotAPI(hostURL, agentId);
    if (messageListener) {
      this.updateListener = messageListener;
    }
  }

  /**
   * It will set the messenger state as open and initialize conversation
   * in case it's needed.
   */
  public toggleMessenger = async (): Promise<void> => {
    if (!this.stateManager.agent.isInitialized) {
      await this.initMessenger();
      await this.stateManager.updateAgent({ isInitialized: true });
    }

    this.stateManager.updateLayout({
      isMessengerOpen: !this.stateManager.layout.isMessengerOpen,
    });

    this.pushUpdate();
  };

  public get isOpen() {
    return this.stateManager.layout.isMessengerOpen;
  }

  public get messages() {
    return this.stateManager.messages;
  }

  public sendMessage = async (message: Message): Promise<void> => {
    try {
      let messagesResponse: APIMessage[] | undefined;
      if (message.type === "jump") {
        messagesResponse = await this.apiClient.postJump(message.label);
      } else {
        messagesResponse = await this.apiClient.postMessage(message.label);
      }
      if (messagesResponse) {
        const messages: Message[] = messagesResponse.map<Message>(messageResponse => ({
          ...messageResponse,
          sender: "bot",
        }));

        this.stateManager.saveMessages(messages);

        this.pushUpdate();
      }
    } catch (err) {
      throw new Error("An error occurred sending a message.");
    }
  };

  /**
   * Notifies the subscriber with the updated messages
   */
  private pushUpdate() {
    if (this.updateListener) {
      this.updateListener();
    }
  }

  private initMessenger = async () => {
    try {
      const messagesResponse = await this.apiClient.postStartConversation();
      if (messagesResponse) {
        const messages: Message[] = messagesResponse.map<Message>(message => ({
          ...message,
          sender: "bot",
        }));

        this.stateManager.saveMessages(messages);

        this.pushUpdate();
      }
    } catch (err) {
      throw new Error("An error occurred initializing messenger.");
    }
  };
}
