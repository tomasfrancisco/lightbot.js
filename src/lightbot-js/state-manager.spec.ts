import { LightbotMessage } from "./messenger";
import { AgentState, LayoutState, StateManager } from "./state-manager";

describe("StateManager", () => {
  describe("localStorage unavailable", () => {
    it("not throw any error on constructor", () => {
      const call = () => new StateManager();
      expect(call).not.toThrow();
    });

    describe("messages", () => {
      let stateManager: StateManager;

      beforeEach(() => {
        stateManager = new StateManager();
      });

      it("adds new message when message is saved", () => {
        const message: LightbotMessage = { sender: "human", label: "test message", type: "plain" };
        stateManager.saveMessages([message]);

        expect(stateManager.messages[0]).toMatchObject(message);
      });

      it("removes message when it pops", () => {
        stateManager.popMessage();

        expect(stateManager.messages[0]).toBeUndefined();
      });
    });

    describe("layout", () => {
      let stateManager: StateManager;
      const initialLayout: LayoutState = { isMessengerOpen: false };
      const overrideLayout: LayoutState = { someValue: true };

      beforeEach(() => {
        stateManager = new StateManager();
        stateManager.updateLayout(initialLayout);
      });

      it("adds layout properties", () => {
        stateManager.updateLayout(overrideLayout);
        expect(stateManager.layout).toEqual({ ...initialLayout, ...overrideLayout });
      });

      it("overrides a property", () => {
        stateManager.updateLayout({ isMessengerOpen: true });
        expect(stateManager.layout.isMessengerOpen).toBeTruthy();
      });
    });
  });

  describe("localstorage available", () => {
    beforeEach(() => {
      require("jest-localstorage-mock");
      localStorage.clear();
    });

    describe("with clean localStorage", () => {
      it("initializes empty messages storage", () => {
        expect(new StateManager().messages).toHaveLength(0);
      });
    });

    describe("constructor when existing localStorage", () => {
      let initialMessages: LightbotMessage[];
      let initialLayout: LayoutState;
      let initialAgent: AgentState;
      let stateManager: StateManager;

      beforeEach(() => {
        initialMessages = [{ type: "plain", sender: "bot", label: "Test message from bot" }];
        initialLayout = { isMessengerOpen: true };
        initialAgent = { isInitialized: true };
        localStorage.setItem(StateManager.keys.messages, JSON.stringify(initialMessages));
        localStorage.setItem(StateManager.keys.layout, JSON.stringify(initialLayout));
        localStorage.setItem(StateManager.keys.agent, JSON.stringify(initialAgent));
        stateManager = new StateManager();
      });

      it("initializes messages from localStorage", () => {
        expect(stateManager.messages).toEqual(initialMessages);
      });

      it("initializes layout from localStorage", () => {
        expect(stateManager.layout).toEqual(initialLayout);
      });

      it("initializes agent from localStorage", () => {
        expect(stateManager.agent).toEqual(initialAgent);
      });
    });

    describe("messages", () => {
      let stateManager: StateManager;

      beforeEach(() => {
        stateManager = new StateManager();
      });

      it("adds new message when message is saved", () => {
        const message: LightbotMessage = { sender: "human", label: "test message", type: "plain" };
        stateManager.saveMessages([message]);

        expect(stateManager.messages[0]).toMatchObject(message);
      });

      it("removes message when it pops", () => {
        stateManager.popMessage();

        expect(stateManager.messages[0]).toBeUndefined();
      });
    });

    describe("layout", () => {
      let stateManager: StateManager;
      const initialLayout: LayoutState = { isMessengerOpen: false };
      const overrideLayout: LayoutState = { someValue: true };

      beforeEach(() => {
        stateManager = new StateManager();
        stateManager.updateLayout(initialLayout);
      });

      it("adds layout properties", () => {
        stateManager.updateLayout(overrideLayout);
        expect(stateManager.layout).toEqual({ ...initialLayout, ...overrideLayout });
      });

      it("overrides a property", () => {
        stateManager.updateLayout({ isMessengerOpen: true });
        expect(stateManager.layout.isMessengerOpen).toBeTruthy();
      });
    });
  });
});
