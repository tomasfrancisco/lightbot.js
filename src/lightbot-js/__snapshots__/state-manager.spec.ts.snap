// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StateManager localStorage unavailable not throw any error on constructor 1`] = `"received is not a function"`;
